LibArchive.Net/LibArchiveReader.cs

```diff
@@ -1,211 +1,268 @@
+        handle = archive_read_new();
+        archive_read_support_filter_all(handle);
+        archive_read_support_format_all(handle);
+        if (archive_read_open_filenames(handle, names.Ptr, (int)blockSize) != 0)
+            Throw();
+    }
+
+    private void Throw()
+    {
+#if !NETSTANDARD2_0
+        throw new ApplicationException(Marshal.PtrToStringUTF8(archive_error_string(handle)));
+#else
+        throw new ApplicationException(PtrToStringUTF8(archive_error_string(handle)));
+#endif
+    }
+
+    protected override bool ReleaseHandle()
+    {
+        return archive_read_free(handle) == 0;
+    }
+
+    public IEnumerable<Entry> Entries()
+    {
+        int r;
+        while ((r=archive_read_next_header(handle, out var entryHandle))==0)
+        {
+            var entry = Entry.Create(entryHandle, handle);
+            if (entry is not null)
+            {
+                yield return entry;
+            }
+        }
+
+        if (r != (int)ARCHIVE_RESULT.ARCHIVE_EOF)
+            Throw();
+    }
+
+    public class Entry
+    {
+        protected readonly IntPtr entryHandle;
+        protected readonly IntPtr archiveHandle;
+
+        public string Name { get; }
+        public EntryType Type;
+        public FileStream Stream => new(archiveHandle);
+
+        public bool IsDirectory => Type == EntryType.Directory;
+        public bool IsRegularFile => Type == EntryType.RegularFile;
+
+        protected Entry(IntPtr entryHandle, IntPtr archiveHandle)
+        {
+            this.entryHandle = entryHandle;
+            this.archiveHandle = archiveHandle;
+#if !NETSTANDARD2_0
+            Name = Marshal.PtrToStringUTF8(archive_entry_pathname(entryHandle)) ?? throw new ApplicationException("Unable to retrieve entry's pathname");
+#else
+            Name = PtrToStringUTF8(archive_entry_pathname(entryHandle)) ?? throw new ApplicationException("Unable to retrieve entry's pathname");
+#endif
+            Type = (EntryType)archive_entry_filetype(entryHandle);
+        }
+
+        internal static Entry? Create(IntPtr entryHandle, IntPtr archiveHandle)
+        {
+            try
+            {
+                return new Entry(entryHandle, archiveHandle);
+            }
+            catch (ApplicationException)
+            {
+                return null;
+            }
+        }
+    }
+
+    public enum EntryType
+    {
+        Directory = 0x4000,  // AE_IFDIR
+        RegularFile = 0x8000 // AE_IFREG
+    }
+
+    public class FileStream : Stream
+    {
+        private readonly IntPtr archiveHandle;
+
+        internal FileStream(IntPtr archiveHandle)
+        {
+            this.archiveHandle = archiveHandle;
+        }
+        
+        public override void Flush()
+        {
+        }
+
+        public override int Read(byte[] buffer, int offset, int count)
+        {
+#if !NETSTANDARD2_0
+            return archive_read_data(archiveHandle, ref MemoryMarshal.GetReference(buffer.AsSpan()[offset..]), count);
```

Comment:
Doesn't `MemoryMarshal.GetReference` just return the first element, so this is equivalent to `ref buffer[offset]`? Which should work on every TFM.